Update ModelManager to support Doctrine\ORM\Query directly

djpretzel opened the pull request:
Makes life easier when you want to set sorting, caching, etc. directly from an admin, e.g. in configureFormFields

<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

I am hitting use cases in admins where I want to change sort order, add caching, etc. which requires modifying the actual Doctrine query object, not the QueryBuilder, and I need to then be able to pass it in. This worked previously in 3.X because there was a default call to execute() which just happened to work on Doctrine's Query class; I think it should be explicitly supported in ModelManager, to make things easier.

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because it does not break BC.

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataDoctrineORMAdminBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->
```markdown
### Added
 - ModelManager::supportsQuery now supports AbstractQuery
 - ModelManager::execute now supports AbstractQuery
```

djpretzel committed:
Update ModelManager to support Doctrine\ORM\Query directly

Makes life easier when you want to set sorting, caching, etc. directly from an admin, e.g. in configureFormFields

VincentLanglet commented on src/Model/ModelManager.php:
```diff
@@ -217,18 +218,19 @@ public function executeQuery(object $query)
         }
 
-        if ($query instanceof ProxyQuery) {
+        if ($query instanceof Query || $query instanceof ProxyQuery) {
```
Please use a different if
```suggestion
        if ($query instanceof Query) {
            return $query->execute();
        }

        if ($query instanceof ProxyQuery) {
```

Because the Query is not returning a Paginator.

VincentLanglet reviewed (commented):
A case could be added for supportsQueryDataProvider
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/tests/Model/ModelManagerTest.php#L141-L154

djpretzel committed:
Update src/Model/ModelManager.php

Co-authored-by: Vincent Langlet <VincentLanglet@users.noreply.github.com>

djpretzel committed:
Update ModelManagerTest.php

Added test for Query class

djpretzel commented:
Committed change & added test - I think it should work?

VincentLanglet commented on tests/Model/ModelManagerTest.php:
```diff
@@ -149,6 +149,7 @@ public function testSupportsQuery(bool $expected, object $object): void
     public function supportsQueryDataProvider(): iterable
     {
         yield [true, new ProxyQuery($this->createMock(QueryBuilder::class))];
+        yield [true, $this->createMock(Query::class)];
```
You forgot `use Doctrine\ORM\Query;`

djpretzel replied:
So I did! I need to get comfortable doing this via IDE instead of the github website, so I get warnings - added!

VincentLanglet replied:
Also with an IDE you can install the vendor and run the tests, seems like it's failing

djpretzel committed:
Update ModelManagerTest.php

Forgot use statement for Query

VincentLanglet commented on tests/Model/ModelManagerTest.php:
```diff
@@ -149,6 +150,7 @@ public function testSupportsQuery(bool $expected, object $object): void
     public function supportsQueryDataProvider(): iterable
     {
         yield [true, new ProxyQuery($this->createMock(QueryBuilder::class))];
+        yield [true, $this->createMock(Query::class)];
```
You have to use
```
        yield [true, new Query()];
```
to avoid the error
```
1) Error
The data provider specified for Sonata\DoctrineORMAdminBundle\Tests\Model\ModelManagerTest::testSupportsQuery is invalid.
PHPUnit\Framework\MockObject\ClassIsFinalException: Class "Doctrine\ORM\Query" is declared "final" and cannot be doubled
/home/runner/work/SonataDoctrineORMAdminBundle/SonataDoctrineORMAdminBundle/tests/Model/ModelManagerTest.php:153
```

djpretzel committed:
Update ModelManagerTest.php

djpretzel committed:
Update ModelManager.php

Switch to AbstractQuery since that's where execute() lives, this way we can support NativeQuery, AND AbstractQuery can be mocked for tests.

djpretzel commented:
Switched to AbstractQuery - supports NativeQuery + easier to mock
